### ISSUE TITLE
Restructure the ompi_proc_t for nspace and scaling

### ISSUE THREAD

**Issue by rhc54** [open-mpi/ompi#7442]
OMPI currently uses an `ompi_proc_t` like this:
```c
struct ompi_proc_t {
    opal_proc_t                     super;

    /* endpoint data */
    void *proc_endpoints[OMPI_PROC_ENDPOINT_TAG_MAX];

    char padding[OMPI_PROC_PADDING_SIZE]; /* for future extensions (OSHMEM uses this area also)*/
};
```
where the `opal_proc_t` consists of:
```c
typedef struct opal_proc_t {
    /** allow proc to be placed on a list */
    opal_list_item_t                super;
    /** this process' name */
    opal_process_name_t             proc_name;
    /** architecture of this process */
    uint32_t                        proc_arch;
    /** flags for this proc */
    opal_hwloc_locality_t           proc_flags;
    /** Base convertor for the proc described by this process */
    struct opal_convertor_t*        proc_convertor;
    /** A pointer to the name of this host - data is
     * actually stored outside of this framework.  */
    char*                           proc_hostname;
} opal_proc_t;
```
The `opal_process_name_t` consists of two 32-bit fields:
```c
typedef uint32_t opal_jobid_t;
typedef uint32_t opal_vpid_t;
typedef struct {
    opal_jobid_t jobid;
    opal_vpid_t vpid;
} opal_process_name_t;
```
If we generalize the `jobid` to conform to the PMIx `pmix_nspace_t`, we would expand the `opal_process_name_t` by an unacceptable amount - the `pmix_nspace_t` is a 256-byte character array. This would consume far too much memory at scale.

What we would ideally like to do is utilize the `pmix_nspace_t` to identify the job while simultaneously reducing the size of the `ompi_proc_t`. This is actually achievable if we:

* reorganize the `ompi_proc_t` to be stored by nspace instead of on a global list
* remove the `opal_process_name_t` from the `opal_proc_t` and replace it with just a `pmix_rank_t`
* adjust the padding of `ompi_proc_t` to retain proper memory alignment

We would definitely _not_ want to put the `ompi_proc_t` objects into a pointer array by rank as this would unnecessarily explode the array on sparsely connected applications. So we would need a clever way of storing them under a new `ompi_nspace_t` object - perhaps in a hash table? I leave it to someone else to design.

Once we have this in place, we can remove the hack in https://github.com/open-mpi/ompi/pull/7441 and just use the `pmix_nspace_t` when referencing a process. This would remove a lot of jobid \<-\> nspace conversions we have to sprinkle around the code, and align us more with the native PMIx integration.

Seems like this might be a good ECP project - would someone from there like to claim it?

**Comment by bosilca** [comment]
What is nspace ?

We cannot replace the opal_process_name_t with pmix_rank_t because an
ompi_proc_t might point to a process from another jobid, and we use the
info from opal_process_name to manipulate the groups.

We can however replace the opal_process_name_t structure with a pointer to
something in PMIx for as long as this is unique per peer.

George.

**Comment by rhc54** [comment]
> What is nspace ?

The namespace is the PMIx equivalent of a jobid. It is a 256-char string array.

> We cannot replace the opal_process_name_t with pmix_rank_t because an
ompi_proc_t might point to a process from another jobid, and we use the
info from opal_process_name to manipulate the groups.

This is why I suggested organizing the `ompi_proc_t` objects according to their namespace. However, it does mean that you would have to reference the namespace somehow so you could link back to discover it from an `ompi_proc_t` - might not be of help.

> We can however replace the opal_process_name_t structure with a pointer to
something in PMIx for as long as this is unique per peer.

We do have the `pmix_proc_t` which is unique per peer. However, it includes that 256-char array in it, which makes it a poor candidate from a memory scaling perspective. I'm looking for a way to avoid duplicating that array for every proc - it would be nice if we only had it in memory once per job and then did everything else based on the rank within that namespace.

**Comment by bosilca** [comment]
I did not say we need to duplicate the pmix_proc_t array. We can simply use
pointers (8 bytes) into your array. This solves the memory issue as the
costly storage is only kept once.

  George.

**Comment by rhc54** [comment]
I'm not communicating this well, so let me try again. If we replace the process name in `ompi_proc_t`with a pointer to a `pmix_proc_t` in PMIx, then all we do is shift the memory problem from OMPI to PMIx. We still wind up with an array of N objects (where N=number of procs) in memory, with each object containing the 256-char array namespace plus the 4-byte rank. If we have 1M procs, that means we have 256Mbytes of namespace data.

I'll grant that pushing it into PMIx does buy us something. If we can share the array of `pmix_proc_t` objects across the procs on a node, then we gain a factor of ppn reduction in overall memory footprint. Extreme scale systems are likely to have 30-100 ppn, so this is a sizable reduction over storing the array inside every proc. However, it is still consuming 256Mbytes of memory on the node, so it remains a significant penalty!

Note that we are trying _not_ to store a `pmix_proc_t` in PMIx memory for every proc for precisely the same reason - it explodes our memory footprint. Even with such a measure, the PMIx footprint for extreme scales is likely too large to be accommodated in most systems, so we are looking at methods for reducing it. Thus, creating an N-object array of `pmix_proc_t` objects in PMIx isn't likely to be an option as we are trying to go the other direction (i.e., reducing our footprint, not growing it).

The large majority of the time, the procs will be from the same namespace. Thus, we could dramatically reduce the memory footprint if we had only one copy of the namespace info. In other words, we would modify `ompi_proc_t` to look like:
```c
pmix_nspace_t *nptr    --> points to pmix_nspace_t for this nspace, could be in PMIx
pmix_rank_t vpid
...
```
The negative to this layout is that the OMPI code would need to reconstruct the full process name if it wanted to use it in some operation - i.e., you would have to construct an object that actually had the namespace+vpid in it. So we would need an accessor function to get the process name. Fortunately, you already introduced such a function earlier - we would just need to modify it (not saying that mod is easy, just pointing out that the need to use an accessor function already has precedence in the code).

Using this approach, the `ompi_proc_t` would grow by 4 bytes (as the 8-byte pointer replaces the 4-byte jobid), but we could handle the namespace without absorbing a much larger impact. I'm not saying this is the best solution, or even a good one - I'm just suggesting that someone might want to begin exploring this issue.

Does this make sense now?

**Comment by bosilca** [comment]
I think my proposal went a step further in the optimizations, because it
does not require a change to the size of the opal_proc_t. Unfortunately, it
appears I also failed to make myself understood.

Let me try again, but first the premises:
- how the storage of the PMIx process tables is done is irrelevant, we
don't need to make any assumptions on how it is done. Personally, I do hope
it was already done at the node level, and the memory used by the PMIx
daemon (or whatever the qualified name is) was mapped into all same-node
processes.
- OMPI layer needs the pmix_name in very few cases: printing a
user-friendly message, dynamic processes (connecting to a process) and
certainly sessions. For all other purposes (which at this point they are
all internal), having an untyped handle (void*) is good enough for as long
as __it is unique for each peer__.

So, If instead of having a tuple (nptr, vpid), we simply change the
definition of opal_process_name_t to be a pointer to a pmix_proc_t (or
however the official name is) and provide accessors (jobid, vpid, and
user-friendly name) then:
- we will not change the size of the opal_proc_t
- we maintain the same level of functionality with minimal changes

Does this make sense to you ?

  George

PS: Just a note regarding the increase in the ompi_proc_t: using your
proposal we will be increasing the size of the ompi_proc_t by 8 bytes due
to alignment constraints on pointers (and the fact that we have an uint16_t
in the middle of the struct).

**Comment by rhc54** [comment]
I'm not sure we can do what you propose. We are not going to store a `pmix_proc_t` for every proc as consuming 256Mbytes is too much. PMIx is going to store a single instance of the namespace array value, and then has info storage that is keyed to the individual ranks. Thus, we don't have a pointer we can give you that directly references the PMIx proc ID (i.e., the `pmix_proc_t`).

We do have internal structures that could perhaps be used in an accessor function to retrieve the corresponding nspace and rank. My concern is that it is hard to define a referencing mechanism that neither requires new PMIx functions just to dereference the provided pointer (which would be difficult to get thru the standards process) nor ties OMPI specifically to the OpenPMIx implementation.

Again, I'm not saying I have an OMPI solution. I just would like to see someone who cares specifically about OMPI investigate the memory footprint problem associated with the `ompi_proc_t`. I'm not sure that pushing this onto PMIx is going to resolve the problem.

**Comment by bosilca** [comment]
The decision to move from a 2-int naming scheme to a 256-byte character array was purely a PMIx decision, taken without any consultation with the communities that currently use PMIx. Thus, the claim that OMPI imposes anything onto PMIx is not only incorrect, but completely reversed.

At the end of the day, OMPI is mostly agnostic to how PMIx name processes internally. What I said is that as long as you can build a bijective function that translates the PMIx name into a 64bits value, OMPI can be adapted to use the new scheme.

**Comment by hppritcha** [comment]
I think @rhc54 has some points here about reorganizing the proc list.  reopening so we don't lose track of this.

**Comment by gpaulsen** [comment]
I put v5.0.x and blocker labels on here, so we remember to come to some consensus on this before we ship v5.0.

**Comment by awlauria** [comment]
I believe this issue can be closed. It was removed from 5.0 tracking. PR merged here:

https://github.com/open-mpi/ompi/pull/7554
